src/common/connection.js

```diff
@@ -176,7 +176,8 @@ class Connection extends EventEmitter {
         this._ws.on('error', error =>
           this.emit('error', 'websocket', error.messsage, error));
         this._ws.on('message', this._onMessage.bind(this));
-        this._onUnexpectedCloseBound = this._onUnexpectedClose.bind(this);
+        this._onUnexpectedCloseBound = this._onUnexpectedClose.bind(this,
+          resolve, reject);
```

Comment:
Hmmm, maybe because it is in the state `closing` when we enter this method. Maybe it would be cleaner to check for the `closing` state and if so, wait for the `close` event before proceeding. To do this I think we would have to extract the `else` block to a separate function. I think it would be a bit simpler because the current code is getting pretty complex.